File: editor/src/main/java/com/raredev/vcspace/editor/completion/CompletionProvider.java
package com.raredev.vcspace.editor.completion;

import com.raredev.vcspace.editor.language.html.completion.HtmlCompletionProvider;
import com.raredev.vcspace.editor.language.java.completion.JavaCompletionProvider;
import com.raredev.vcspace.editor.language.kotlin.completion.KotlinCompletionProvider;
import com.raredev.vcspace.editor.language.lua.completion.LuaCompletionProvider;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class CompletionProvider {

  private static final Map<Class<?>, CompletionProvider> providers = new HashMap<>();

  public abstract List<VCSpaceCompletionItem> getCompletions(CompletionParams params);

  public static void registerCompletionProviders() {
    providers.put(HtmlCompletionProvider.class, new HtmlCompletionProvider());
    providers.put(JavaCompletionProvider.class, new JavaCompletionProvider());
    providers.put(KotlinCompletionProvider.class, new KotlinCompletionProvider());
    providers.put(LuaCompletionProvider.class, new LuaCompletionProvider());
  }

  public static CompletionProvider getCompletionProvider(Class<?> clss) {
    return providers.get(clss);
  }
}

File: feature/editor/src/main/java/com/teixeira/vcspace/editor/completion/CompletionProvider.kt
/*
 * This file is part of Visual Code Space.
 *
 * Visual Code Space is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Visual Code Space is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Visual Code Space.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira.vcspace.editor.completion

abstract class CompletionProvider {
  abstract fun getCompletions(params: CompletionParams?): List<VCSpaceCompletionItem?>?

  companion object {
    private val providers: Map<Class<*>, CompletionProvider> = HashMap()

    fun registerCompletionProviders() {
      if (providers.isEmpty()) {}
    }

    fun getCompletionProvider(clss: Class<*>): CompletionProvider? {
      return providers[clss]
    }
  }
}